llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13176,6 +13176,61 @@ static SDValue performXORCombine(SDNode *N, SelectionDAG &DAG,
   return combineSelectAndUseCommutative(N, DAG, /*AllOnes*/ false, Subtarget);
 }
 
+// Look for (abs (sub (zext X), (zext Y))).
+// Rewrite as (zext (sub (zext (max X, Y), (min X, Y)))) if the user is an add
```

Comment:
Can't the sub be done at the narrower type as well?  (a >=u b) should imply that (a-b) doesn't underflow, and thus the high bits are always zero?

Reply:
It can, but it would leave a bare vzext.vf2 later. I was trying to carefully create a widening sub and a widening add or widening reduction to minimize the number of individual vector operations.

Reply:
For the workload in question we can pull that vext.vf2 through through the chain of adds that sums the absolute difference pieces. We could use an i16 accumulator for the beginning of the chain and switch to an i32 accumulator later in the chain.

Naive use of the computeKnownBits could get us some of that to prove the overflows don't happen. Need to check the l length of the chain in the workload to see if that would exceed computeKnownBits depth limit.

Reply:
(This is mostly a response to myself)

Here's the alive2 proof for the transformation in this patch:

https://alive2.llvm.org/ce/z/XoCBZ5

Note the need for noundef on the source parameters.  Alternatively, we could use freeze in the target.

Here's my proposed variant:

https://alive2.llvm.org/ce/z/f7MdJe

Reply:
Is lowering (abs (sub X, Y)) to (sub (umax x, y), (umin x, y)) worthwhile doing on its own, ignoring pulling through the zexts for now?

Reply:
> Is lowering (abs (sub X, Y)) to (sub (umax x, y), (umin x, y)) worthwhile doing on its own, ignoring pulling through the zexts for now?

I'm not sure that's valid.

Reply:
I was thinking that we would check the sub doesn't overflow with computeKnownBits, e.g. that the upper bits of X and Y are zero: https://alive2.llvm.org/ce/z/MZuw8V